### PR TITLE
feat(callback): add scalyr callback plugin

### PR DESCRIPTION
**Pull request by martinm82**
##### SUMMARY
Introducing a new callback plugin to push task/playbook results to Scalyr. The module is heavily inspired by the sumologic and splunk callbacks.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
scalyr

##### ADDITIONAL INFORMATION

TODO:
- [ ] Add unit tests
- [x] Check how API keys could be used in Ansible vault instead in ansible.cfg or as env vars

### CODE REVIEW

File: plugins/callback/scalyr.py [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+DOCUMENTATION = r'''
+callback: scalyr
+type: aggregate
+short_description: Sends task result events to Scalyr
+author: Martin Migasiewicz (@martinm82)
+version_added: 1.3.0
+description:
+  - This callback plugin will send task results as JSON formatted events to Scalyr.
+requirements:
+  - Whitelisting this callback plugin
```
Comment (felixfontein) [inline plugins/callback/scalyr.py:19]:
```suggestion
  - Whitelisting this callback plugin.
```
Sentences in `description` should always be complete, i.e. end with some kind of punctuation.

File: plugins/callback/scalyr.py [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+  - 'Create a Log Access Write Key in Scalyr (https://app.eu.scalyr.com/keys)'
+  - 'Define the Scalyr API URL and API key in ansible.cfg'
+  - 'Based on the splunk and sumologic callbacks'
+options:
+  url:
+    description: URL to the Scalyr API endpoint
```
Comment (felixfontein) [inline plugins/callback/scalyr.py:25]:
```suggestion
    description: URL to the Scalyr API endpoint.
```

File: plugins/callback/scalyr.py [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+    env:
+      - name: SCALYR_AUTHTOKEN
+    ini:
+      - section: callback_scalyr
+        key: authtoken
+    type: string
+'''
+
+EXAMPLES = '''
+examples: >
```
Comment (felixfontein) [inline plugins/callback/scalyr.py:44]:
```suggestion
examples: |
```
Or do you really want to remove newlines?

File: plugins/callback/scalyr.py [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+                {
+                    "ts": str(int(secondsToNanoSeconds(time.time()))) ,
+                    "type": 0,
+                    "sev": 3,
+                    "attrs": {
+                        "uuid": result._task._uuid,
+                        "status": state,
+                        "runtime": runtime,
+                        "user": self.user,
+                        "ansible_version": ansible_version,
```
Comment (felixfontein) [inline plugins/callback/scalyr.py:117]: Please note that this is not the Ansible version, but the ansible-base / ansible-core version.

**Review by felixfontein — COMMENTED** [review]
Also please note that all new plugins need tests.

File: plugins/callback/scalyr.py (hunk 3 above)
Comment (martinm82) [reply plugins/callback/scalyr.py:44]: Copy&paste. I haven't thought about what the difference is.

File: plugins/callback/scalyr.py (hunk 4 above)
Comment (martinm82) [reply plugins/callback/scalyr.py:117]: Is there any other way to get the ansible version=

**Review by martinm82 — COMMENTED** [review]
> Also please note that all new plugins need tests.

Yes, definitely. I forgot to add some todos

File: plugins/callback/scalyr.py (hunk 4 above)
Comment (felixfontein) [reply plugins/callback/scalyr.py:118]: There's no way, actually. You can only get hold of the ansible-base/ansible-core version, which happens to be the Ansible version for Ansible 2.9.

**Comment by martinm82** [comment]
@felixfontein are there any guidelines how to create integration tests for callbacks? I haven't found any useful so far.

**Comment by felixfontein** [comment]
The tests for a callback called `scalyr` must be in `tests/integration/targets/callback_scalyr`. In `tests/integration/targets/` you can see tests for two other callbacks.

Since your callback talks to a server, it will be more complicated. Maybe the best approach is to create a mini Python HTTP server which acts close enough to a real Scalyr instance. From your it looks like you simply have to accept POSTs for one URL, and accept a JSON payload. The server could do some basic validation whether the messages look like you expect them.

**Comment by felixfontein** [comment]
You could also write a sanitized version (i.e. with normalized hostnames, timestamps etc. that could differ from run to run) of the message to disk, like a concatenation of pretty-printed JSON objects, and then compare the resulting file with an expected copy by `diff`, and react on the `diff` return value. That way, one can quickly see why the tests fails (you see the diff to the expected version).

**Comment by felixfontein** [comment]
@martinm82 are you still working on this?

needs_info

**Comment by martinm82** [comment]
@felixfontein I haven't worked on it since a while but I will for sure come back to it and implement the missing tests.

**Comment by felixfontein** [comment]
@martinm82 regular reminder ping :)

**Comment by felixfontein** [comment]
needs_info

**Comment by felixfontein** [comment]
Closing since there was no reaction.

**Pull request closed without merging**